Update how to (#6)
AgPipeline/template-rgb-plot: HOW_TO.md

1. [Setup](#setup): Click the `Use this template` button in GitHub to make a copy of this repository (or run `git clone`)
2. [Definitions](#definitions): Fill in and modify the definitions in the algorithm_rgb.py file
3. [Algorithm](#algorithm): Replace the code in the `calculate` function with your algorithm
4. [Generate](#generate): Run `generate.py` to create a Dockerfile
5. [Test](#test): Run the `testing.py` script to run your algorithm and validate the results
6. [Docker](#build_docker): Create a Docker image for your algorithm and publish it

> (Optional) 7. [Testing Your docker image](#test_docker):

Reviewer: This is not rendering properly

Be sure to save your changes.

### Generate the docker build command file <a name="generate" />

> It's time to make a Docker image so that it can run as part of a workflow.

Reviewer: ```suggestion
It's time to generate the Dockerfile that's used to build Docker images.
Docker images can be used as part of a workflow.
```

Be sure to save your changes.

> ### Generate the docker build command file <a name="generate" />

Reviewer: ```suggestion
### Generate the Docker build command file <a name="generate" />
```

If there are other dependencies needed by your algorithm, add them to the `packages.txt` file.
The packages listed will be installed using `apt-get` as part of the Docker build process.

### Test your algorithm <a name="test"/>
A testing script named `testing.py` is provided for testing your algorithm.
It checks whether the configuration is correct for testing the files by making sure that the arguments in algorithm_rgb as well as the image files are in the correct format
What isn't provided in the template repository are the plot-level RGB images to test against.
It's expected that you will either provide the images or use a standard set that can be downloaded from [Google Drive](https://drive.google.com/file/d/1xWRU0YgK3Y9aUy5TdRxj14gmjLlozGxo/view?usp=sharing).

> The testing script requires `numpy` and `gdal` to be installed on the testing system.

Reviewer: This should probably be moved after the 'For example...' line below

For example, if your files reside in `/user/myself/test_images` the command to test could be the following:
```./testing.py /user/myself/test_images```

The testing script expects to have either a list of source plot image files, or a folder name, or both specified on the command line.

### Create the Docker image <a name="build_docker" />

> Now that you have generated your `Dockerfile` and specified any Python modules and other packages needed by your algorithm, you are ready to create a Docker image of your algorithm.

Reviewer: ```suggestion
Now that you have generated your `Dockerfile` as described [above](#generate) and specified any Python modules and other packages needed by your algorithm, you are ready to create a Docker image of your algorithm.
```

### Test your algorithm <a name="test"/>
A testing script named `testing.py` is provided for testing your algorithm.
It checks whether the configuration is correct for testing the files by making sure that the arguments in algorithm_rgb as well as the image files are in the correct format

The testing script requires `numpy` and `gdal` to be installed on the testing system.

If your files reside in `/user/myself/test_images` the command to test could be the following:
```./testing.py /user/myself/test_images```

What isn't provided in the template repository are the plot-level RGB images to test against.
It's expected that you will either provide the images or use a standard set that can be downloaded from [Google Drive](https://drive.google.com/file/d/1xWRU0YgK3Y9aUy5TdRxj14gmjLlozGxo/view?usp=sharing).

> The testing script expects to have either a list of source plot image files, or a folder name, or both specified on the command line.

Reviewer: What happened to running testing.py from within a docker container? I thought that it was useful, otherwise I would have suggested that it be removed.

### Create the Docker image <a name="build_docker" />
Now that you have generated your `Dockerfile` as described [above](#generate) and specified any Python modules and other packages needed by your algorithm, you are ready to create a Docker image of your algorithm.

A sample Docker build command could be: ```docker build -t my_algorithm:latest ./```
Please refer to the Docker documentation for additional information on building a docker image.

Once the image is built, you can run it locally or push it to an image repository, such as [DockerHub](https://hub.docker.com/).
Please note that there may be naming requirements for pushing images to a repository.

> ### (OPTIONAL) Producing the Docker Image <a name="produce_docker" />

Reviewer: Isn't this testing and not producing?

```./testing.py /user/myself/test_images```

What isn't provided in the template repository are the plot-level RGB images to test against.
It's expected that you will either provide the images or use a standard set that can be downloaded from [Google Drive](https://drive.google.com/file/d/1xWRU0YgK3Y9aUy5TdRxj14gmjLlozGxo/view?usp=sharing).

The testing script expects to have either a list of source plot image files, or a folder name, or both specified on the command line.

### Create the Docker image <a name="build_docker" />
Now that you have generated your `Dockerfile` as described [above](#generate) and specified any Python modules and other packages needed by your algorithm, you are ready to create a Docker image of your algorithm.

A sample Docker build command could be: ```docker build -t my_algorithm:latest ./```
Please refer to the Docker documentation for additional information on building a docker image.

Once the image is built, you can run it locally or push it to an image repository, such as [DockerHub](https://hub.docker.com/).
Please note that there may be naming requirements for pushing images to a repository.

> ### (OPTIONAL) Using Docker to run testing.py <a name="test_docker">

Reviewer: ```suggestion
#### (OPTIONAL) Using Docker to run testing.py <a name="test_docker">
```

- `docker run` tells Docker to run an instance of the image (specified later in the command) (Refer to [docker run](https://docs.docker.com/engine/reference/run/) documentation)
- `--rm` tells Docker to remove the container (an image instance) when it's completed
- `it` allows you to have a stdin stream and terminal driver added to the docker container allowing an interactive session
- `-v "pwd":/mnt` bind mounts a volume to the docker container so that the current working directory (given by pwd) will be copied into the volume
- `--entrypoint /mnt/testing.py` defines the Docker container that will be run, with testing.py mounted to that container 
- `my_algorithm:latest` is the image to run (the running image is known as a *container*)
- `/mnt/images` mounts the sample plot images to the running docker container

Output should be in the format of image name and calculated value for that image on a single line for each of the images in the images folder.
Example output from the images in the [Google Drive](https://drive.google.com/file/d/1xWRU0YgK3Y9aUy5TdRxj14gmjLlozGxo/view?usp=sharing) 
is contained below for plot images folder, which is titled sample_plot_images: 

> Filename,size of image channels -  (pixels),

Reviewer: ```suggestion
```Filename,size of image channels -  (pixels),
```

/mnt/sample_plot_images/rgb_17_7_W.tif,7000
/mnt/sample_plot_images/rgb_40_11_W.tif,7000
/mnt/sample_plot_images/rgb_6_1_E.tif,7000
/mnt/sample_plot_images/rgb_1_2_E.tif,7000
/mnt/sample_plot_images/rgb_33_8_W.tif,7000
/mnt/sample_plot_images/rgb_5_11_W.tif,7000

> 

Reviewer: \`\`\`
needs to be added if above suggestion is accepted

> ### (OPTIONAL) Testing Image Production <a name="production" />

Reviewer: ```suggestion
#### (OPTIONAL) Production Testing of Image  <a name="production" />
```

Please note that there may be naming requirements for pushing images to a repository.

> **Testing the Docker image**

Reviewer: A better title for this might be 'Testing production readiness of Docker image'; hopefully something shorter. This section should be moved below '(OPTIONAL)...' below

using the same image setup as used when testing your algorithm, a sample command line to run the image could be:

> ```docker run --rm --mount "src=/user/myself,target=/mnt,type=bind" my_algorithm:latest --working_space "/mnt" --metadata "mnt/08f445ef-b8f9-421a-acf1-8b8c206c1bb8_metadata_cleaned.json "/mnt/images"```

Reviewer: This is **still** using TERRA REF metadata files. These aren't supported for drones

- `my_algorithm:latest` is the image to run (the running image is known as a *container*)
- `--working_space "/mnt"` lets the software in the container know where its working disk space is located; files are created here
- `--metadata "mnt/08f445ef-b8f9-421a-acf1-8b8c206c1bb8_metadata_cleaned.json"` specifies that the metadata file
08f445ef-b8f9-421a-acf1-8b8c206c1bb8_metadata_cleaned.json will be made available to the container
- `"/mnt/images"` specifies where the plot-level image files are located

The `--mount` command line parameter is important since it allows the running container to access the local file system.
The container can then load the images from the file system directly, without having to perform any copies.
The parameters after the Docker image name are all relative to the target folder specified with this command line parameter.

Once the image files have been processed, the resulting CSV file(s) will be located in the folder at `/user/myself` (in this example).

The result.json file should tell you what errors were found in the checks from testing.py (make sure to check the output in the CSV file(s) 
even if the result.json file does not find errors)

> ### (OPTIONAL) Testing the Docker container: 

Reviewer: A better title might be: `(Optional) Testing your algorithm using a Docker container`

> ### (OPTIONAL) Testing the Docker container: 

Reviewer: I think a better title would be something along the lines of: 'Using Docker to run testing.py' or something like that; a more accurate description of what this section contains

> Now you have built your docker image. In order to run the testing.py file from within a container, do the following command:

Reviewer: I would get rid of the first sentence.

> 

Reviewer: Should refer them to the docker documentation on the `run` command.

> ```docker run --rm -it -v `pwd`:/mnt --entrypoint bash my_algorithm:latest```

Reviewer: Wouldn't a better docker command call the testing.py file directly instead of running a bash shell?

Options different from those above include:

> - `-it` allows you to have a stdin stream and terminal driver added to the docker container

Reviewer: ```suggestion
- `-it` allows you to have a stdin stream and terminal driver added to the docker container allowing an interactive session
```

> - `v` bind mounts a volume to the docker container

Reviewer: Elaborate by adding the full `-v` parameter values and explaining them

Once the image is built, you can run it locally or push it to an image repository, such as [DockerHub](https://hub.docker.com/).
Please note that there may be naming requirements for pushing images to a repository.

> ### (OPTIONAL) Testing your Docker Image <a name="test_docker">

Reviewer: ```suggestion
### (OPTIONAL) Using Docker Image to test the algorithm <a name="test_docker">
```

- `/mnt/images` mounts the sample plot images to the running docker container

Output should be in the format of image name and calculated value for that image on a single line for each of the images

> in the images folder. Example output from the images in the [Google Drive](https://drive.google.com/file/d/1xWRU0YgK3Y9aUy5TdRxj14gmjLlozGxo/view?usp=sharing) 

Reviewer: Our standard for MD files, which probably isn't documented anywhere, is to have sentences on individual lines
